### PR TITLE
Fix bugs when run on CPU 64 and add output debug of each layer

### PR DESCRIPTION
**Pull request by parai**
Here is a demo notebook show how to run nnom on ubuntu amd64 https://github.com/autoas/CMSIS_5/blob/c8c5e6bd7eb0a7b809c6769a82b03817843e7e79/CMSIS/NN/Examples/ARM/arm_nn_examples/mnist/mnist-on-keras.ipynb.
During the phase, fix some bugs for 64 bit CPU and add output debug of each layer.
And another thing modified is about the bias_shift, the original code is "layer->bias_shift = w->shift - b->shift", this really makes me confused, why subtraction? According to my understanding, I changed it, and it works.
Also add a SConsript that to be used to integrated with https://github.com/autoas/as or rt-thread.

### CODE REVIEW

**Comment by majianjia** [comment]
You did a great job!
It's a pleasure to see this lib can run on 64-bit platforms with only a little modification.

Here are a few comments on your PR, please prove me wrong if you think so. 

---

For the `const`  that added to the weights and bias,
I didn't add it because I was thinking to make the weights and biases trainable in the future. Even 
Is there any way around to avoid making them const?
However, let make them const by the meantime. Change it later when i figure out how to train these quantised weights. 

---

To answer you question of  `layer->bias_shift = w->shift - b->shift` . 

`layer->bias_shift = w->shift - b->shift` is because my scripts output the original fraction part of the quantised bias, which is not as the same meaning of its name.  
`#define BIAS_SHIFT  5` means the bias is Q2.5
In the [quantise guid](https://github.com/ARM-software/ML-KWS-for-MCU/blob/master/Deployment/Quant_guide.md#quantizing-kws-models) they mentioned:

> Using quant_test.py, we get the quantized weights are of the format Q0.7 (i.e. in range -1,1) and the quantized biases have 8 bits for fractional point (i.e. range -0.5,0.5). So, the product (layer_input x weights) will have 9 bits in the fractional part (Q5.2 x Q0.7 = Qx.9) and biases need to be shifted left by 1 to get to the same representation (i.e. Qx.8<<1 = Qy.9). 

Which means that the bias shift taken by CMSIS-NN is the difference between the product of (input x weight) and bias. 
`bias_shift = (weight_shift + input_shift) - bias_shift`
so the complete version is 
`layer->bias_shift = w->shift + layer->in->shift - b->shift`
In the previous code,` layer->in->shift` is always 0 since it is only using Q7.0. 
This is a problem that i haven't fixed. 

[My answer on the issue in CMSIS-NN](https://github.com/ARM-software/CMSIS_5/issues/541#issuecomment-468255380) shows how to calculate bias_shift and output_shift. 

In short, they should be calculated by these equations:
`output_shift =  output_n - (input_n + weight_n) `
`bias_shift =   (input_n + weight_n) - bias_n `
The the output_shift for CMSIS-NN is fixed to right shift. The value needs to be inverted `output_shift = - output_shift` 

---

As you can see, this is an incomplete job. 
`the layer->in->shift` will never know until the previous layer is compiled (or the previous layer._output_shape() is called.)

Thus the calculations of these arguments (output_shift and bias_shift for every layer that potentially change these formats) must be put in the layer._output_shape(). 

The next urgent TODO is to find out how to pass the output Qmn for some layer (the layers accept an output_shift argument, such as convolution, dense). To be noticed, not all of them can change the Qm.n of their output. 

Though i am currently busy with my PhD thesis, i couldn't do more coding before the end of March.
If you leave them, i will fix them later with the idea below. 

My initial idea is to add output_shift for the layer into it's  nnom_weight_t
Because 
1. Only those layers required weights can change their output Qmn. 
2. These shifts can be generated with weights together, will not necessary to pass to the Layer APIs manually. 

Please tell me what do you think.
Thank you very much.

File: src/nnom_layers.c [hunk 1]
```diff
@@ -117,7 +117,7 @@ nnom_layer_t *Conv2D(uint32_t filters, nnom_shape_t k, nnom_shape_t s, nnom_padd
 	layer->bias = b;
 	layer->weights = w;
 	layer->output_shift = w->shift;
-	layer->bias_shift = w->shift - b->shift; // bias is quantized to have maximum shift of weights
```
Comment (majianjia) [inline src/nnom_layers.c]: The bias_shift is the difference between a product of (input x weight), which is (input_n + weight_n) in Qmn format.
 
I will rather keep it as it is now. 
This is related to a TODO of supporting different Qmn format across layers. 
Please see my [comment](https://github.com/majianjia/nnom/pull/7#issuecomment-468775756) for more explanation.

**Comment by majianjia** [comment]
BTW, could you commit your notebook as an example into the /examples folder?
It will be very convincing as a demo to have the quantised model validated on PC instead of on MCU. 
I will modify the README.MD according to your works
Thank you

File: src/nnom_layers.c (hunk 1 above)
Comment (parai) [reply src/nnom_layers.c]: I can somehow understanding your explain, for each of the layer, I know that if its input in format Qim.in, and its output format is Qom.on, and the weights in format Qwm.wn and bias in format Qbm.Qbn, then there is below relation ship:
bias_shift = wn+in-bn
output_shift = wn+in - on
And by the way, I think it's not reasonable to fix each layer's input data to format Q7.0, as my study mnist notebook shows that its not fixed. For the input, it's Q0.7, for th conv1, it's Q2.5,..etc.

If possible, we can arrange a meeting to talk about it.

**Comment by parai** [comment]
Hi @majianjia :
Hope things going good for your PhD thesis.
I somehow still think for an inference framework, it's not necessary to make things the same as keras, as keras is a framework for training and research. I think the future of NNOM is to make it good for mode deployment and to do optimization to do inference quickly, to improve performance, to reduce memory usage, etc.
And for the shift, I think we can generated by scripts directly, and for the compiling phase of NNOM, somehow, I think its not necessary.  Also as you mentioned before, if scripts can generated the model directly like the CMSIS NN CIFAR10 example, the NNOM has no meaning, but I am not agree, NNOM can be a high level representation of model, make things easy, as the original ARM CMSIS NN is low level and difficult for new beginners, and its really need a complicated scripts to do the model generation. But with NNOM, I think it could make the scripts easy to generate a model.

**Comment by majianjia** [comment]
Thanks
I see your point, let discuss a bit more in the meeting before merging to the master.

**Comment by majianjia** [comment]
Its good to talk with you.
Minutes:
Using (unsigned long) for address pointer to compatible with 64bit. 
output_shift, bias_shift are planned to generate from script directly. 
Layer names not shown in the tested 64-bit platform.

**Comment by parai** [comment]
Hi:
I am looking into the mnist densenet example and code/scripts, I have few questions:
1. could you please provide the relate C code which construct the mnist densenet model on nnom?
2. For layer add/sub/concate, especially for concate as frequently used, there is no consideration of the inputs that with different Q formats. That would not a problem as currently layer output has been fixed to Q7.0 as told by you.
3.You force the bias Q format somehow the same as kernel weight format, that may result accuracy loss.

**Comment by majianjia** [comment]
Hi Parai,
I have added the codes for mcu as your request, please check
The accuracy is not tested, it might be worst as the current NNoM support only one output shift. 

For the add/sub/mult, they are using CMSIS-DSP, which does not support changing the output_shifting. 
The concat is one necessary layer for complex structures (Inception, DenseNet..) which is not provided by CMSIS-NN, so I have to write our own version. it simply copies the data by the selected axis. 

This is one of the reasons which was preventing me to implement multiple output_shifting. 
Let's take the Inception as an example,
Conv layer (2.1) can change the output_shift, while the MaxPool layer (2.2) cannot. 
Then the merging layer (3) has to deal with different input format. 
image

All add/sub/mult and concat in NNoM have 2 or more input memory. So they also called 'merge method', which are passed to `model.merge((method(), in1, in2 ...))` as input. 
Currently, cannot change the input buffers' Qmn format. 

There are 2 solutions that i am currently thinking. 
1. In NNoM side, simply add an invisible layer or modify the current merge method to shift the input data. 
2. In PC side (script), makes all the parallel layers output the same format. 

Neither way is perfect. 

Again, this is related to the TODO job for supporting multiple output_shift.
Please tell me what do you think.

**Comment by parai** [comment]
Hi @majianjia :
I somehow prefer method 2. I somehow think for this situation that from training side to get a method(such as fake_quant) to force those inputs of add/concat/sub  to has the same Q Format.

Again, thanks for your very detailed reply.

**Comment by majianjia** [comment]
Hi @parai 
I also think method 2 is great, too. it won't affect the performance on NNoM side.
If one model can use fake_quant, it is easier to set all relative layers to the same Qmn.

Just need to be careful, when we want to write a script is trying to identify a pre-trained model (trained without quantisation) automatically. The script somehow has to identify these parallel links in the structures. I think it is doable but I am not very confident about that.

**Comment by parai** [comment]
Hi:
I am working on that and I have already have done an alpha version that could resolve the Q format to the same for concat/sub,etc. But there is accuracy issue,  I will check with my original mnist cnn demo firstly.
And why would you prefer to use densenet, I think densenet is not good as need to set Q format the same, and not that easy to train a workable model.
